File: cplusplus/p144.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(vector<int>& ret, TreeNode* root) {
        ret.push_back(root->val);
        if (root->left != NULL) dfs(ret, root->left);
        if (root->right != NULL) dfs(ret, root->right);
    }
    
    vector<int> preorderTraversal(TreeNode* root) {
        vector<int> ret;
        if (root == NULL) {
            return ret;
        } else {
            dfs(ret, root);
            return ret;
        }
    }
};

File: python/2023_daily/p144.py
from typing import Optional, List


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def _preorderTraversal(self, root: Optional[TreeNode]):
        self.container.append(root.val)
        if root.left is not None:
            self._preorderTraversal(root.left)
        if root.right is not None:
            self._preorderTraversal(root.right)

    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        self.container = []
        if root is not None:
            self._preorderTraversal(root)
        return self.container
